modules/apps/data-engine/data-engine-taglib/src/main/resources/META-INF/resources/data_layout_builder/js/components/rule-builder/Calculator/Calculator.es.js

```diff
@@ -0,0 +1,267 @@
+		disableOperators = true;
+		showOnlyRepeatableFields = true;
+	}
+
+	if (
+		tokens.length === 0 ||
+		(tokens.length > 0 && tokens[tokens.length - 1].type !== Token.LITERAL)
+	) {
+		disableDot = true;
+	}
+
+	if (
+		tokens.length > 0 &&
+		tokens[tokens.length - 1].type === Token.OPERATOR
+	) {
+		disableOperators = true;
+	}
+
+	return {
+		disableDot,
+		disableFunctions,
+		disableNumbers,
+		disableOperators,
+		showOnlyRepeatableFields,
+	};
+}
+
+function shouldAddImplicitMultiplication(tokens, newToken) {
+	const lastToken = tokens[tokens.length - 1];
+
+	return (
+		lastToken !== undefined &&
+		((newToken.type === Token.LEFT_PARENTHESIS &&
+			lastToken.type !== Token.OPERATOR &&
+			lastToken.type !== Token.FUNCTION &&
+			lastToken.type !== Token.LEFT_PARENTHESIS) ||
+			(newToken.type === Token.FUNCTION &&
+				lastToken.type !== Token.OPERATOR &&
+				lastToken.type !== Token.LEFT_PARENTHESIS) ||
+			(newToken.type === Token.VARIABLE &&
+				(lastToken.type === Token.VARIABLE ||
+					lastToken.type === Token.LITERAL)) ||
+			(newToken.type === Token.LITERAL &&
+				(lastToken.type === Token.VARIABLE ||
+					lastToken.type === Token.FUNCTION)))
+	);
+}
+
+function addTokenToExpression({expression, tokenType, tokenValue}) {
+	const newToken = new Token(tokenType, tokenValue);
+	const tokens = Tokenizer.tokenize(expression);
+
+	if (shouldAddImplicitMultiplication(tokens, newToken)) {
+		tokens.push(new Token(Token.OPERATOR, '*'));
+	}
+
+	tokens.push(newToken);
+
+	return Tokenizer.stringifyTokens(tokens);
+}
+
+function removeTokenFromExpression({expression}) {
+	const tokens = Tokenizer.tokenize(expression);
+
+	const removedToken = tokens.pop();
+
+	if (
+		removedToken &&
+		removedToken.type === Token.LEFT_PARENTHESIS &&
+		tokens.length &&
+		tokens[tokens.length - 1].type === Token.FUNCTION
+	) {
+		tokens.pop();
+	}
+
+	return Tokenizer.stringifyTokens(tokens);
+}
+
+const Calculator = forwardRef(
+	(
+		{
+			expression: initialExpression = '',
+			fields = [],
+			functions = [],
+			index = 0,
+			onEditExpression = () => {},
+			repeatableFields = getRepeatableFields(fields),
+			resultSelected,
+			placeholder = Liferay.Language.get(
+				'the-expression-will-be-displayed-here'
+			),
+		},
+		ref
+	) => {
+		const [expression, setExpression] = useStateSafe(initialExpression);
+
+		const {
+			disableDot,
+			disableFunctions,
+			disableNumbers,
+			disableOperators,
+			showOnlyRepeatableFields,
+		} = useMemo(() => getStateBasedOnExpression(expression), [expression]);
+
+		const dropdownItems = showOnlyRepeatableFields
+			? repeatableFields
+			: fields;
+
+		const updateExpression = useCallback(
+			({index, newExpression}) => {
+				setExpression(newExpression);
+
+				onEditExpression({
+					index,
+					newExpression,
+				});
+			},
+			[onEditExpression, setExpression]
+		);
+
+		useEffect(() => {
+			const newExpression = expression.replace(/[[\]]/g, '');
+
+			updateExpression({index, newExpression});
```

Comment:
I was trying to reproduce the same behaviour as here:

https://github.com/liferay/liferay-portal/blob/68abe7cbe0b7e197aaad4c81408bd929fdecb2f1/modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/Calculator/Calculator.es.js#L101


It looks like a string mask for the expression. Do you have a suggestion for replacing it?

Reply:
Hmm, I don't know what use case for this, it could be for a specific fix that they did so maybe the best thing now would be to move this to the `updateExpression` function.

Reply:
Done, It looks like a macaroni with cheese